libsrc/grabber/drm/DRMFrameGrabber.cpp

```diff
@@ -0,0 +1,490 @@
+	}
+
+	return rc;
+}
+
+int DRMFrameGrabber::grabFrame(Image<ColorRgb> & image)
+{
+	bool rc (true);
+
+	if (_isEnabled && !_isDeviceInError)
+	{
+		if ( getScreenInfo() )
+		{
+			for(auto & framebuffer : framebuffers)
+			{
+				_pixelFormat = GetPixelFormatForDrmFormat(framebuffer.second->pixel_format);
+				uint64_t modifier = framebuffer.second->modifier;
+				int fb_dmafd = 0;
+
+				DebugIf(verbose, _log, "Framebuffer ID: %d - Width: %d - Height: %d  - DRM Format: %c%c%c%c - PixelFormat: %s"
+					, framebuffer.first // framebuffer ID
+					, framebuffer.second->width // width
+					, framebuffer.second->height // height
+					, framebuffer.second->pixel_format         & 0xff
+					, (framebuffer.second->pixel_format >> 8)  & 0xff
+					, (framebuffer.second->pixel_format >> 16) & 0xff
+					, (framebuffer.second->pixel_format >> 24) & 0xff
+					, QSTRING_CSTR(pixelFormatToString(_pixelFormat))
+				);
+
+				if (_pixelFormat != PixelFormat::NO_CHANGE && modifier == DRM_FORMAT_MOD_LINEAR)
+				{
+					int w = framebuffer.second->width;
+					int h = framebuffer.second->height;
+					Grabber::setWidthHeight(w, h);
+
+					int size;
+					int ret = drmPrimeHandleToFD(_deviceFd, framebuffer.second->handles[0], O_RDONLY, &fb_dmafd);
+					if (ret < 0)
+					{
+						continue;
+					}
+
+					uint8_t *mmapFrameBuffer;
+					if (_pixelFormat == PixelFormat::I420 || _pixelFormat == PixelFormat::NV12)
+					{
+						size = (w * h * 3) / 2;
+						mmapFrameBuffer = (uint8_t*)mmap(nullptr, size, PROT_READ, MAP_PRIVATE, fb_dmafd, 0);
+						if (mmapFrameBuffer != MAP_FAILED)
+						{
+							_imageResampler.processImage(mmapFrameBuffer, w, h, w, _pixelFormat, image);
+							munmap(mmapFrameBuffer, size);
+							close(fb_dmafd);
+							break;
+						}
+						else
+						{
+							Error(_log, "Format: %c%c%c%c failed. Error: %s"
+								, framebuffer.second->pixel_format         & 0xff
+								, (framebuffer.second->pixel_format >> 8)  & 0xff
+								, (framebuffer.second->pixel_format >> 16) & 0xff
+								, (framebuffer.second->pixel_format >> 24) & 0xff
+								, strerror(errno)
+							);
+							break;
+						}
+					}
+					else if (_pixelFormat == PixelFormat::RGB32 || _pixelFormat == PixelFormat::BGR32)
+					{
+						size = w * h * 4;
+						mmapFrameBuffer = (uint8_t*)mmap(nullptr, size, PROT_READ, MAP_PRIVATE, fb_dmafd, 0);
+						if (mmapFrameBuffer != MAP_FAILED)
+						{
+							_imageResampler.processImage(mmapFrameBuffer, w, h, w * 4, _pixelFormat, image);
+							munmap(mmapFrameBuffer, size);
+							close(fb_dmafd);
+							break;
+						}
+						else
+						{
+							Error(_log, "Format: %c%c%c%c failed. Error: %s"
+								, framebuffer.second->pixel_format         & 0xff
+								, (framebuffer.second->pixel_format >> 8)  & 0xff
+								, (framebuffer.second->pixel_format >> 16) & 0xff
+								, (framebuffer.second->pixel_format >> 24) & 0xff
+								, strerror(errno)
+							);
+							break;
+						}
+					}
+
```

Comment:
There is a lot of code duplication here. Despite the size and line length (which depends on fourcc) everything else seems identical